src/2025h2/reflection-and-comptime.md

```diff
@@ -0,0 +1,158 @@
+
+So for this experimental impl we would do
+
+```rust
+#[compile_time_only]
+const fn type_of(id: TypeId) -> &'static Type;
+```
+
+These functions can't be run at runtime, because that would require there to be some global table somewhere that maps all `TypeId`s to their repr. This is an obvious no-go in my book.
+
+an demonstration impl (absolutely not salvageable for anything that could be landed!) can be found [here](https://github.com/rust-lang/rust/compare/master...oli-obk:rust:compile-time-reflection)
+
+### Why not continue where uwuflection left off?
+
+See https://soasis.org/posts/a-mirror-for-rust-a-plan-for-generic-compile-time-introspection-in-rust/ for details on what uwuflection is
+
+
+#### Structural processing
+
+it makes procedural processing of type information very hard. E.g. to get the 3rd element of a tuple you need to 
+
+```rust
+<introwospect_type::<YourType> as FieldDescriptor<3>>
+```
+
+so to compute that index you need a constant. you can't just use a for loop to iterate over the indices.
+
+we chose const fn over associated consts and generics which could compute the same thing as the const fn, just more expensively and mostly purely functional. so going to assoc consts again seems like it goes against that
+
+This can somewhat be resolved by adding a compile-time for loop feature to the language, that will just expand the body N times in a macro/loop-unrolling kind of way.
+
+#### generic const exprs
+
+in order to use uwuflection in types in generic code you need to either write infallible code with min const generics or you need to add uwuflection bounds (Lots, think typenum) , which defeats the purpose.
+
+### Why not go full zig-style comptime?
+
+zig's approach to comptime from a very high level is effectively
+
+* generate AST for all source files
+* pick the `main` function and start compiling it and looking for what it needs to be compiled
+* if a comptime function call is found, look only for what code that needs to compile, compile it and produce the resulting code of the comptime function
+* continue the main compilation, which may now invoke the generated code and start compiling that
+
+we do not experiment with this approach at this time, because the compiler is not set up in a way to permit proc macros from accessing type information from the current crate.
+While there are ongoing refactorings that go into the direction of potentially allowing more of that in the future, that future seems to be more than 5 years away at my best guess.
+
+* the compiler is not set up to perform codegen while type information is already available. It possibly never will, and it would be an immense amount of work to get there. I'm doing lots of refactorings that would need to be done for sth like that anyway, even if the goal is just better incremental and general compilar architecture.
```

Comment:
> the compiler is not set up to perform codegen while type information is already available

I am confused by this statement. We have type information during codegen. Is there a negation missing here somewhere?

Reply:
```suggestion
* the compiler is not set up to add AST nodes while type information is already available. It possibly never will, and it would be an immense amount of work to get there. I'm doing lots of refactorings that would need to be done for sth like that anyway, even if the goal is just better incremental and general compilar architecture.
```
?

Reply:
Yea I wanna edit everything with the feedback given on monday, but this seems like a good start